tests/testthat/test-exercise.R

```diff
@@ -0,0 +1,226 @@
+})
+
+test_that("render_exercise() returns envir_result up to error", {
+  exercise <- mock_exercise(
+    user_code = c("y <- 2", "stop('boom')", "z <- 3"),
+    chunks = list(
+      mock_chunk("setup-1", "x <- 1")
+    ),
+    setup_label = "setup-1",
+    error_check = "unevaluated, triggers error_check in render_exercise()"
+  )
+
+  exercise_result <- withr::with_tempdir(render_exercise(exercise, new.env()))
+
+  # the error during render causes a checker evaluation, so we can recover
+  # the environments from the checker_args returned by the debug checker
+  exercise_result <- exercise_result$feedback$checker_args
+
+  expect_s3_class(exercise_result$last_value, "simpleError")
+  expect_equal(conditionMessage(exercise_result$last_value), "boom")
+
+  expect_false(identical(exercise_result$envir_prep, exercise_result$envir_result))
+  expect_setequal(ls(exercise_result$envir_prep), "x")
+  expect_setequal(ls(exercise_result$envir_result), c("x", "y"))
+  expect_identical(get("y", exercise_result$envir_result), 2)
+})
+
+test_that("render_exercise() with errors and no checker returns exercise result error", {
+  exercise <- mock_exercise(
+    user_code = "stop('user')",
+    chunks = list(mock_chunk("setup-1", "stop('setup')")),
+    setup_label = "setup-1"
+  )
+
+  exercise_result <- withr::with_tempdir(render_exercise(exercise, new.env()))
+  expect_s3_class(exercise_result, "learnr_exercise_result")
+  expect_identical(exercise_result$error_message, "setup")
+  expect_null(exercise_result$feedback)
+
+  exercise <- mock_exercise(user_code = "stop('user')")
+  exercise_result <- withr::with_tempdir(render_exercise(exercise, new.env()))
+  expect_s3_class(exercise_result, "learnr_exercise_result")
+  expect_identical(exercise_result$error_message, "user")
+  expect_null(exercise_result$feedback)
+})
+
+# evaluate_exercise() -----------------------------------------------------
+
+test_that("serialized exercises produce equivalent evaluate_exercise() results", {
+  exercise <- mock_exercise(
+    user_code = c("z <- 3", "z"),
+    chunks = list(
+      mock_chunk("setup-1", "x <- 1"),
+      mock_chunk("setup-2", "y <- 2", exercise.setup = "setup-1")
+    ),
+    setup_label = "setup-2",
+    global_setup = "w <- 0",
+    check = I("identical(eval(parse(text = 'w + x + y + z'), envir_result), 6)")
+  )
+
+  # From internal_external_evaluator() in R/evaluators.R
+  exercise_serialized <- jsonlite::toJSON(exercise, auto_unbox = TRUE, null = "null")
+
+  ex_eval_local <- evaluate_exercise(exercise, new.env(), TRUE)
+
+  rmt <- callr::r_bg(
+    function(exercise) {
+      exercise <- jsonlite::fromJSON(exercise, simplifyDataFrame = FALSE)
+      # AsIs attribute doesn't survive serialization, but it's only used for testing
+      exercise$check <- I(exercise$check)
+      learnr:::evaluate_exercise(exercise, new.env(), TRUE)
+    },
+    args = list(exercise = exercise_serialized)
+  )
+  rmt$wait()
+  ex_eval_rmote <- rmt$get_result()
```

Comment:
Using `:::` assumes that the local learnr and installed learnr are the same and will lead to funny issues if pkg versions disagree.

I suggest we remove the `callr` trip and just serialize, parse, and check again (all locally).

```suggestion
  ## Get a _local_ evaluation
  ex_eval_local <- evaluate_exercise(exercise, new.env(), TRUE)

  ## Mock an _external_ evaluation
  # From internal_external_evaluator() in R/evaluators.R
  exercise_serialized <- jsonlite::toJSON(exercise, auto_unbox = TRUE, null = "null")
  exercise_unserialized <- jsonlite::fromJSON(exercise, simplifyDataFrame = FALSE)
  # AsIs attribute doesn't survive serialization, but it's only used for testing
  exercise_unserialized$check <- I(exercise_unserialized$check)
  ex_eval_rmote <- evaluate_exercise(exercise, new.env(), TRUE)
```